### PR TITLE
Add Hessian check for 2-parameter problems

### LINKED ISSUE #362: Add a simple post-optimisation identifiability check
### Feature description

Add a function that can assess the identifiability of an optimisation result and return an informative message to the user.

### Motivation

To help users consider the identifiability of an optimal point.

### Possible implementation

Compute a finite difference approximation of the Hessian matrix around the result of an optimisation and classify the eigenvalues.

### Additional context

Useful webpage related to this issue: https://sethna.lassp.cornell.edu/Sloppy/WhatAreSloppyModels.html

### PR DESCRIPTION
**Pull request by NicolaCourtier**
# Description

Add a `classify_using_Hessian` function that can assess the identifiability of an optimisation result and return an informative message to the user.

## Issue reference
Fixes #362.

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

**Comment by NicolaCourtier** [comment]
I think we should continue with the slightly low coverage (98%) - the Hessian check is missing a test on the case where the optimal point is a saddle point for which we would need a relevant optimisation problem (this relates to #391).

**Review by BradyPlanden — COMMENTED** [review]
Thanks @NicolaCourtier, a few suggestions/comments

File: pybop/_classification.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+import numpy as np
+
+
+def classify_using_Hessian(optim, x=None, epsilon=1e-2):
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_classification.py:4]: It requires 18 cost evaluations and an eigenvalue calculation, so less than half a second on my old laptop. How does that compare?

File: pybop/_classification.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        # Estimate the Hessian using second-order accurate central finite differences
+        # cfd_hessian = np.zeros((2, 2))
+        # cfd_hessian[0, 0] = costs[2,1,0] - 2 * costs[1,1,0] + costs[0,1,0]
+        # cfd_hessian[0, 1] = (costs[2,2,0] - costs[2,0,0] + costs[0,0,0] - costs[0,2,0]) / 4
+        # cfd_hessian[1, 0] = cfd_hessian[0, 1]
+        # cfd_hessian[1, 1] = costs[1,2,0] - 2 * costs[1,1,0] + costs[1,0,0]
+
+        # Estimate the Hessian using fourth-order accurate central finite differences
+        cfd_hessian = np.zeros((2, 2))
+        cfd_hessian[0, 0] = (
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/_classification.py:82]: I think this is a good idea! But perhaps not necessary for this feature addition.

File: pybop/_classification.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
+from typing import Optional
+
+import numpy as np
+
+from pybop import OptimisationResult
+
+
+def classify_using_Hessian(
```
Comment (BradyPlanden) [inline pybop/_classification.py:8]: A bit pedantic, but in line with Pep8 standards:
```suggestion
def classify_using_hessian(
```

**Review by BradyPlanden — APPROVED** [review]
LGTM - just one suggestion.

**Pull request merged**